### ISSUE TITLE
statusline has bug with unicode emoji

### ISSUE THREAD

**Issue by ttys3** [neovim/neovim#14224]
This bug confused me for a rather long time, until today I finally got the time to dig into it.

- `nvim --version`:
```
NVIM v0.5.0-dev+1178-gf0ace6d41
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla
 -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/ttys3/.config/nvim/neovim/build/co
nfig -I/home/ttys3/.config/nvim/neovim/src -I/home/ttys3/.config/nvim/neovim/.deps/usr/include -I/usr/include -I/home/ttys3/.config/nvim/neovim/build/src/nvim/auto -I/home/ttys3/.config/nvim/neovim/build/include
Compiled by ttys3@ttys3

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"

Run :checkhealth for more info
```

- Operating system/version: `Ubuntu 20.10 x86_64`
- 
- Terminal name/version:

both `GNOME Terminal 3.38.0 using VTE 0.62.0 +BIDI +GNUTLS +ICU  +SYSTEMD` 
and `alacritty 0;7.2` 

- `$TERM`: `xterm-256color`

### Steps to reproduce using `nvim -u NORC`

minimal rc:  `statusline_issue.vim`

```vim
set nu
set updatetime=100

lua << EOF
math.randomseed(os.time())
EOF

function! AleStatus() abort
    let ok = 'ok'
        let prefix =  '🇦 '
    let num = luaeval('math.random(1,2)')
    if num == 1 
        return prefix
    else    
        return prefix.  ok
    endif
endfunction

set statusline=%=%{AleStatus()}
set statusline+=[0x%B]

" 0123456789 

```

```
nvim -u ./statusline_issue.vim ./statusline_issue.vim

# move chars from `0` to `9`, and you'll find the wrong code shows up.
```

### Actual behaviour

char hex randomly show chars `0xXX` and `0xXXX` (yes, three XXX chars)

we all know that for `4`, its ASCII code is `0x34`, but neovim shows as `0x343`, and the `]` been eaten by neovim

![neovim-statusline-unicode-emoji-Peek 2021-03-27 16-25](https://user-images.githubusercontent.com/41882455/112715091-5d778380-8f19-11eb-94f2-be7dbddc8b3f.gif)

![image](https://user-images.githubusercontent.com/41882455/112714731-b396f780-8f16-11eb-8926-e5e4d31610a5.png)

### Expected behavior

char hex works ok, should show chars (`0xXX`)

### Misc

we use 0 to 9 because their ASCII code is very easy to remember.
# so that we can found the problem very easily

use `0x%B` is just for example.

the actual screenshot in my case will like this:

![image](https://user-images.githubusercontent.com/41882455/112714980-3b313600-8f18-11eb-8515-227fca58a8ba.png)

and this:

![image](https://user-images.githubusercontent.com/41882455/112714993-569c4100-8f18-11eb-8f13-07342d1bbca0.png)

> the char `🇦` codepoints is U+1F1E6 <https://emojipedia.org/emoji/%F0%9F%87%A6/>

> if we replace it with another emoji, say Melon `🍈` <https://emojipedia.org/melon/>, everything works fine. see below gif

**Comment by ttys3** [comment]
![st-emoji-another-ok-Peek 2021-03-27 16-32](https://user-images.githubusercontent.com/41882455/112715296-45543400-8f1a-11eb-8ac0-37c3b960c145.gif)

**Comment by jdhao** [comment]
Can reproduce this issue on macOS with Alacritty.

kitty on macOS does not have this issue. Maybe it is a terminal issue?

**Comment by ttys3** [comment]
I think it should not related to terminal.

I tried with Gnome Terminal, also reproduced

**Comment by zeertzjq** [comment]
May be another character width issue.

**Comment by zeertzjq** [comment]
Likely fixed by #19686
